SIPS/sip-20.md

```diff
@@ -1,191 +1,298 @@
-- `source`: The URL that the message was sent from. Snaps can use this to differentiate between different endpoints. This SHOULD be the exact same URL as specified in manifest.
+type GetWebSocketsResult = WebSocketConnection[];
+```
 
-The specification of types of `onExternalData` and its parameters:
+An example of usage is given below.
 
 ```typescript
-/**
- * A text message received from a WebSocket.
- *
- * @property type - The type of the message.
- * @property message - The message as a string.
- */
-type WebSocketTextMessage = {
-  type: 'text';
-  message: string;
+snap.request({
+  method: "snap_getWebSockets",
+})
+```
+
+
+### Handling WebSocket Events
+
+Snaps can handle WebSocket events by implementing the `onWebSocketEvent` handler. This handler will be called whenever a WebSocket event occurs, such as receiving a message, opening a connection, closing a connection, or encountering an error.
+
+```typescript
+import { OnWebSocketEventHandler } from "@metamask/snap-sdk";
+
+export const onWebSocketEvent: OnWebSocketEventHandler = async ({ event }) => {
+  switch (event.type) {
+    case "message":
+      // Handle incoming message
+      console.log(`Message received from ${event.origin}:`, event.data);
+      break;
+    case "open":
+      // Handle connection opened
+      console.log(`WebSocket connection opened with ID ${event.id} from ${event.origin}`);
+      break;
+    case "close":
+      // Handle connection closed
+      console.log(`WebSocket connection closed with ID ${event.id} from ${event.origin}`);
+      break;
+    case "error":
+      // Handle error
+      console.error(`WebSocket error occurred with ID ${event.id} from ${event.origin}`);
+      break;
+  }
 };
+```
 
-/**
- * A binary message received from a WebSocket.
- *
- * @property type - The type of the message.
- * @property message - The message as an ArrayBuffer.
- */
-type WebSocketBinaryMessage = {
-  type: 'binary';
-  message: ArrayBuffer;
+```typescript
+type OnWebSocketEventArgs = {
+  event: WebSocketEvent;
 };
+```
+
+```typescript
+type OnWebSocketEventResponse = void;
+```
+
+#### Event Types
 
-/**
- * A message received from a WebSocket.
- *
- * @property type - The type of the message.
- * @property message - The message as a string or an ArrayBuffer, depending on
- * the type.
- */
-type WebSocketData = {
-  type: 'websocket';
-  data: WebSocketTextMessage | WebSocketBinaryMessage;
+the type for an `onWebSocketEvent` handler function's arguments. 
+
+```typescript
+export type WebSocketEvent =
+  | WebSocketMessage
+  | WebSocketOpenEvent
+  | WebSocketCloseEvent
+  | WebSocketErrorEvent;
+```
+
+##### Message Event
+
+This event is triggered when a message is received over a WebSocket connection. The message can be either text or binary data.
+
+```typescript
+export type WebSocketTextMessage = {
+  type: "text";
+  message: string;
+};
+export type WebSocketBinaryMessage = {
+  type: "binary";
+  message: number[];
```

Comment:
Wondering if this should be a base64 string, since we use that in other places for binary data.

Reply:
@FrederikBolding

Reply:
What would be the reasoning? When I proposed this I was thinking we wanted to mirror the browser WebSocket API as much as possible

Reply:
Just consistency with other APIs. Even better if we could have a translation layer on the execution side turning it into `Uint8Array` before passing to the Snap, but 🤷